### PR TITLE
[SkyServe] Add service schema and change to new service YAML

### PR DESCRIPTION
**Pull request by cblmemo**
This PR adds a schema checking for our newer version service YAML, and change to new YAML config pattern.

An example:

```yaml
service:
  port: 8082
  readiness_probe:
    path: /health
    timeout: 1200
  replica_policy:
    min_replica: 1
    max_replica: 5             # optional
    qpm_upper_threshold: 10    # optional, if qpm is higher, scale up
    qpm_lower_threshold: 1     # optional, if qpm is lower, scale down
```

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/serve/autoscalers.py [hunk 1]
```diff
@@ -95,10 +97,11 @@ def __init__(self,
         :param kwargs:
         """
         super().__init__(*args, **kwargs)
-        self.query_interval = query_interval
-        self.upper_threshold = upper_threshold
-        self.lower_threshold = lower_threshold
         self.min_nodes = min_nodes
+        self.max_nodes = max_nodes or min_nodes
```
Comment (infwinston) [inline sky/serve/autoscalers.py:101]: should we check `min_nodes <= max_nodes`?

File: sky/serve/autoscalers.py [hunk 2]
```diff
@@ -78,10 +80,10 @@ class RequestRateAutoscaler(Autoscaler):
 
     def __init__(self,
                  *args,
-                 query_interval: int = 10,
-                 upper_threshold: int = 10,
-                 lower_threshold: int = 2,
                  min_nodes: int = 1,
+                 max_nodes: Optional[int] = 10,
+                 upper_threshold: Optional[int] = 1,
+                 lower_threshold: Optional[int] = 0,
```
Comment (infwinston) [inline sky/serve/autoscalers.py:86]: should we set the default of three arguments to be `None` given the default policy will be fixed #replica.

File: sky/serve/autoscalers.py [hunk 3]
```diff
@@ -95,10 +97,11 @@ def __init__(self,
...
         """
         super().__init__(*args, **kwargs)
-        self.query_interval = query_interval
-        self.upper_threshold = upper_threshold
-        self.lower_threshold = lower_threshold
         self.min_nodes = min_nodes
+        self.max_nodes = max_nodes or min_nodes
+        self.query_interval = 60  # Therefore thresholds represent queries per minute.
+        self.upper_threshold = upper_threshold or 1
+        self.lower_threshold = lower_threshold or 0
```
Comment (infwinston) [inline sky/serve/autoscalers.py:104]: we should remove `or 1` and `or 0` right?

File: sky/serve/autoscalers.py [hunk 4]
```diff
@@ -137,8 +140,9 @@ def evaluate_scaling(self):
             self.scale_up(1)
             self.last_scale_operation = current_time
         elif requests_per_node > self.upper_threshold:
```
Comment (infwinston) [inline sky/serve/autoscalers.py:142]: what if `self.upper_threshold` is `None`, then we shouldn't run the below code right?
we probably should refactor the code a bit to incorporate fix-node policy and autoscaling policy.

File: sky/serve/common.py [hunk 5]
```diff
@@ -1,32 +1,75 @@
...
+                                      'Invalid service YAML:')
+
+        service_config = {}
+        service_config['readiness_path'] = config['readiness_probe']['path']
+        service_config['readiness_timeout'] = config['readiness_probe']['timeout']
+        service_config['app_port'] = config['port']
+        service_config['min_replica'] = config['replica_policy']['min_replica']
+        service_config['max_replica'] = config['replica_policy'].get('max_replica', None)
+        service_config['qpm_upper_threshold'] = config['replica_policy'].get('qpm_upper_threshold', None)
+        service_config['qpm_lower_threshold'] = config['replica_policy'].get('qpm_lower_threshold', None)
```
Comment (infwinston) [inline sky/serve/common.py:45]: do we want to use `qps` as discussed?

File: sky/serve/examples/http_server/task.yaml [hunk 6]
```diff
@@ -9,4 +9,9 @@ run: python3 server.py
 
 service:
   port: 8081
-  readiness_probe: /health
+  readiness_probe:
+    path: /health
+    timeout: 12000
```
Comment (infwinston) [inline sky/serve/examples/http_server/task.yaml:14]: let's call it `readiness_timeout`?

File: sky/serve/load_balancers.py [hunk 7]
```diff
@@ -12,12 +12,13 @@
 
 class LoadBalancer:
 
-    def __init__(self, infra_provider, endpoint_path, post_data=None):
+    def __init__(self, infra_provider, endpoint_path, timeout, post_data=None):
```
Comment (infwinston) [inline sky/serve/load_balancers.py:15]: let's call it `readiness_timeout`?

File: sky/utils/schemas.py [hunk 8]
```diff
@@ -143,6 +143,52 @@ def get_storage_schema():
...
+                'additionalProperties': False,
+                'properties': {
+                    'min_replica': {
+                        'type': 'integer',
+                    },
+                    'max_replica': {
+                        'type': 'integer',
+                    },
+                    'qpm_upper_threshold': {
+                        'type': 'integer',
```
Comment (infwinston) [inline sky/utils/schemas.py:181]: should we accept `float`?

**Review by infwinston — COMMENTED** [review]
Thanks @cblmemo ! The code looks great and I left a few comments.

File: sky/serve/autoscalers.py (hunk 4 above)
Comment (cblmemo) [reply sky/serve/autoscalers.py:142]: Sure.. I'll just check if it is `None` for now, then if user only specify `upper_threshold`, that means user only want to autoscale up and #replica will only increase until reach `max_replica`; vise versa for only specify `lower_threshold` case. Only if user specified both upper and lower threshold will enable auto sale up/down. We might need a discussion on here.

File: sky/serve/common.py (hunk 5 above)
Comment (cblmemo) [reply sky/serve/common.py]: Resolved in #2276.

**Comment by cblmemo** [comment]
Close and move to #2276.

**Pull request closed without merging**